woocommerce/woocommerce
# [Experimental] Product Filters: Overlay interaction using server context

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes #51868, replaces https://github.com/woocommerce/woocommerce/pull/52550.

This PR uses a simple approach for overlay interaction: instead of holding the navigation until the apply button is pressed like #52550, this PR doesn't limit the navigation at all, meaning every time shoppers change filters, a navigation is triggered which updates the URL and page content. Doing so comes with props and cons:

#### Pros:
- Simple implementation: We just need to keep the `isOverlayOpened` context unchanged between navigations to make filtering inside the overlay work like on the desktop.
- Fast filtering: As we aren't holding the navigation, by the time shoppers click on the Apply button, the navigation event was triggered for a while, which helps return the results faster.

#### Cons:
- It makes more request to server, but given that it's the same behavior on desktop, I don't consider it a con.


#### Notes

- As we're rely on navigation, the UI may be lag depend on how fast the network is. Selecting a filter may take a noticeable delay before it appears in the Active block. It's not the specific issue in overlay but the desktop version has it too. We need a state management on the client side to do optimistic update, which deserve an issue and several PRs.
- Clicking on the close button doesn't restore the previously selected filters. I think it's more beneficial and less confused to shopper than restoring previous selected filters. Big stores are doing the same, like hm.com or adidas.com.
 


### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.
1. Add `Product Filters` to the `Product Catalog` template.
2. Ensure there is at least a filter using Chips and one using Checkbox List.
3. Save and go to the frontend.
4. Toggle mobile emulator or resize browser to less than 600px width.
5. See overlay toggle button.
6. Click on the button open the overlay.
7. Select some filters, see the filters are selected and appears in the Active block.
8. Uncheck some filters, see they got removed from the Active block, too.
9. Remove some filters from the Active block, see the corresponding filters are unchecked.
10. Click on the Apply button, see the overlay closes and the Product Collection updated.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
